// fats-command-server/fats-command-server.js
/*
 * fats server command
 * */
exports.name = 'server';
exports.usage = '<command> [options]';
exports.desc = 'launch a web server';
exports.register = function (commander) {

	var fats = require('fats-kernel');
	var serverRoot = process.cwd();
	var ip = require('ip').address();
	var socket;

	var daemonErrMsg = {
		'8' : 'port is already in use'
	};

	commander
		.option('-w, --watch <switch>', 'on/off, automatic refresh browsers', String, 'off')
		.option('-r, --root <root>', 'document root', String, serverRoot)
		.option('-p, --port <port>', 'server listen port', 3000)
		.option('-n, --hostname <hostname>', 'server use hostname', String, '')
		.action(function () {

			var webServerName = 'fats server';
			var path = require('path');
			var fs = require('fs');
			var args = Array.prototype.slice.call(arguments);
			var options = args.pop();
			var cmd = args.shift();
			var root = options.root || serverRoot;
			var port = parseInt(options.port);
			var watch = options.watch;
			var hostname = options.hostname != '' ? options.hostname : ip;

			socket = require('socket.io-client')('http://localhost:' + port);

			var daemon = require('daemonize2').setup({
				main: path.join(__dirname, './server.js'),
				name: webServerName,
				silent: true,
				pidfile: 'fatsDaemon.pid',
				argv: ['--root', root, '--port', port, '--watch', watch, '--hostname', hostname]
			});

			// startEvent
			var startEvent = function () {
				var pid = daemon.status();
				if (pid) {
					var serverCfg = fats.util.readJSON(__dirname + '/server.json');
					fats.log.warning(webServerName + ' is already running on ' + serverCfg.hostname + ':' + serverCfg.port);
					process.exit(0);
				} else {

					var config = {
							root: root,
							port: port,
							watch: watch,
							hostname: hostname || ip
						},
						cfgStr = JSON.stringify(config);

					daemon.start(function (err, pid) {

						if (err) {
							fats.log.error(daemonErrMsg[err.message] || 'unkonw error');
							process.exit(0);
						} else {

							/*
							 * write server config to file
							 * */
							fs.writeFileSync(__dirname + '/server.json', cfgStr);

							if ('on' == watch) {
								watchEvent(root);
							}

							/*
							 * open browser
							 * */
							var host;
							if (hostname.length == 0) {
								host = ip;
							} else {
								host = hostname;
							}

							var msg = webServerName + ' is running on ' + host + ':' + config.port;
							fats.log.info(msg.bold);

							fats.util.openURL('http://' + host + ':' + port);

							if ('on' != watch) {
								setTimeout(function () {
									process.exit(0);
								}, 1000);
							}
						}

					});
				}
			};

			// restartEvent
			var restartEvent = function () {
				var serverCfg = fats.util.readJSON(__dirname + '/server.json'), daemon;

				daemon = require('daemonize2').setup({
					main: path.join(__dirname, './server.js'),
					name: webServerName,
					silent: true,
					pidfile: 'fatsDaemon.pid',
					argv: ['--root', root || serverCfg.root, '--port', port || serverCfg.port, '--watch', watch || serverCfg.watch, '--hostname', hostname || serverCfg.hostname]
				});

				daemon.stop(function (err) {
					daemon.start(function (err, pid) {
						if (err) {
							fats.log.error(daemonErrMsg[err.message] || 'unkonw error');
							process.exit(0);
						} else {
							var msg = webServerName + ' is restart on ' + hostname + ':' + port;
							fats.log.info(msg.bold);

							var config = {
									root: root,
									port: port,
									watch: watch,
									hostname: hostname || ip
								},
								cfgStr = JSON.stringify(config);
							fs.writeFileSync(__dirname + '/server.json', cfgStr);

							if ('on' == watch) {
								watchEvent(root);
							} else {
								process.exit(0);
							}
						}
					});

				});
			};

			// watch dir
			var timer = null;
			var watchEvent = function (dir) {
				fs.watch(dir, function (event, filename) {
					if (filename) {
						clearTimeout(timer);
						timer = setTimeout(function () {
							// reload browser
							socket.emit('message', 'reload');
						}, 1000);
					}
				});
			};

			switch (cmd) {
				case 'start':
					startEvent();
					break;

				case 'restart':
					restartEvent();
					break;

				case 'status':
					var pid = daemon.status(), msg;
					if (pid) {
						var serverCfg = fats.util.readJSON(__dirname + '/server.json');
						msg = webServerName + ' is running on ' + serverCfg.hostname + ':' + serverCfg.port;
					} else {
						msg = webServerName + ' server is not started'
					}
					fats.log.info(msg.bold);
					process.exit(0);
					break;

				case 'stop':
					daemon.stop(function () {
						fats.log.info(webServerName + ' stoped');
						process.exit(0);
					});
					break;

				case 'kill':
					daemon.kill();
					break;

				default:
					commander.help();
					return;

			}

		}).on('--help', function () {
			console.log('  Examples:');
			console.log('');
			console.log('    $ fats server start');
			console.log('    $ fats server start -w off -p 3000');
			console.log('    $ fats server stop');
			console.log('');
		});

	commander
		.command('start')
		.description('start server');

	commander
		.command('stop')
		.description('shutdown server');

	commander
		.command('restart')
		.description('restart server');

};
