// fats-command-init/fats-command-init.js
/*
 * fats init command
 * */
require("date-format-lite");
var fs = require('fs');
var path = require('path');
exports.name = 'init';
exports.usage = '[options]';
exports.desc = 'initialization project by template';
exports.register = function (commander) {

	var projectName, hasBranches, template,
		fats = require('fats-kernel'),
		root = process.cwd(),
		inquirer = require('inquirer');

	var yesOrNoArr = ['yes', 'no'];
	var templateArr = ['default'], tplDir = path.dirname(path.dirname(require.main.filename)) + '/lib/templates';

	//read custom template
	fs.readdir(tplDir, function (err, paths) {
		paths.forEach(function (path) {
			var _src = tplDir + '/' + path;
			fs.stat(_src, function (err, st) {
				if (err) {
					throw err;
				}
				if (st.isDirectory() && !fats.util.inArray(path, templateArr)) {
					templateArr.push(path);
				}
			});
		})
	});

	// init custom info
	function initPrompt() {
		inquirer.prompt([
			{
				type: 'input',
				name: 'name',
				default: 'demo',
				message: 'Project name',
				validate: function (val) {
					var done = this.async();
					var targetDir = root + '/' + val;
					var exists = fs.existsSync(targetDir);
					// Do async stuff
					setTimeout(function () {
						if (exists) {
							// Pass the return value in the done callback
							done("Folder already exists, input again");
							return;
						}
						// Pass the return value in the done callback
						done(true);
					}, 600);
				}
			},
			{
				type: 'input',
				name: 'branches',
				choices: yesOrNoArr,
				default: 'no',
				validate: function (val) {
					return fats.util.inArray(val, yesOrNoArr);
				},
				message: 'Has branches [' + yesOrNoArr.join('/') + ']?'
			},
			{
				type: 'input',
				name: 'template',
				choices: templateArr,
				default: 'default',
				validate: function (val) {
					return fats.util.inArray(val, templateArr);
				},
				message: 'Template'
			}
		], function (opts) {
			fats.util.extend(opts, {root: root});
			initAction(opts);
		})
	}

	// init action
	function initAction(opts) {
		checkDir(opts);
		addProject(opts);
		writePackage(opts);
	}

	commander.action(function () {
		initPrompt();
	});

	// checkDir
	function checkDir(opts) {
		var targetDir = root + '/' + opts.name;

		// check dir exists
		var exists = fs.existsSync(targetDir);
		if (!exists) {
			fs.mkdirSync(targetDir);
		}
	}

	//write package
	function writePackage(opts) {

		//write packagejson
		var packageJsonXml =
			'{\n'
				+ '\t"name": '
				+ '"' + opts.name + '"'
				+ ",\n"
				+ '\t"branches": '
				+ '"' + opts.branches + '"'
				+ ",\n"
				+ '\t"template": '
				+ '"' + opts.template + '"'
				+ '\n'
				+ '}\n';

		fs.writeFile(root + '/' + opts.name + '/package.json', packageJsonXml, function (err) {
			if (err)
				throw err;
		});

	}

	//path add project
	function addProject(opts) {
		var thisPath = path.dirname(path.dirname(process.argv[1])) + '/lib/templates/' + opts.template;
		var acTionPath = opts.root + "/" + opts.name;
		var appinfo = fats.util.readJSON(path.dirname(path.dirname(__dirname)) + '/package.json');
		// copy files
		var data = {
			name: opts.name,
			APP_NAME: appinfo.name,
			APP_VERSION: appinfo.version,
			CREATE_TIME: new Date().format('UTC:YYYY-MM-DD hh:mm:ss'),
			CREATE_DATE: new Date().format('UTC:YYYY.MM.DD')
		};
		fats.util.copyDir(thisPath, acTionPath, data);
		console.log('Init success!');
	}
};
