// fats-command-init/fats-command-init.js
/*
 * fats init command
 * */
exports.name = 'init';
exports.usage = '[options]';
exports.desc = 'initialization project by template';
exports.register = function (commander) {

	var projectName, hasBranches, template,
		fats = require('fats-kernel'),
		path = process.cwd(),
		inquirer = require('inquirer');

	var yesOrNoArr = ['yes', 'no'];
	var templateArr = ['fats-default', 'xxxxxx', 'yyyyyyyyyyyy', 'eeeeeeeeeeeeee'];

	// 初始化确认信息
	function initPrompt() {
		inquirer.prompt([
			{
				type: 'input',
				name: 'projectName',
				default: 'Unnamed',
				message: 'Project name：'
			},
			{
				type: 'input',
				name: 'hasBranches',
				choices: yesOrNoArr,
				default: 'no',
				validate: function (val) {
					return fats.util.inArray(val, yesOrNoArr);
				},
				message: 'Has branches [' + yesOrNoArr.join('/') + ']：'
			},
			{
				type: 'input',
				name: 'template',
				choices: templateArr,
				default: 'fats-default',
				validate: function (val) {
					return fats.util.inArray(val, templateArr);
				},
				message: 'Template：'
			}
		], function (opts) {
			fats.util.extend(opts, {path: path});
			initAction(opts);
		})
	}

	// 初始化实际动作
	function initAction(opts) {
		console.log(opts);
	}

	commander.action(function () {
		initPrompt();
	});

};
